### DISCUSSION TITLE
Emotion SWC plugin label options don’t work for css prop, [filename] segment erroneously includes file extension

### DISCUSSION THREAD

**Discussion by GrantASL19** [vercel/next.js#44952] [Ideas]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

Operating System:
  Platform: darwin
  Arch: x64
  Version: Darwin Kernel Version 21.4.0: Fri Mar 18 00:45:05 PDT 2022; root:xnu-8020.101.4~15/RELEASE_X86_64
Binaries:
  Node: 16.6.2
  npm: 7.20.3
  Yarn: 1.22.11
  pnpm: N/A
Relevant packages:
  next: 12.1.6-canary.5
  react: 17.0.2
  react-dom: 17.0.2

### Describe the Bug

The Emotion SWC plugin label generation doesn’t match `@emotion/babel-plugin` in two ways:

1. `className`s added via the `css` prop don’t include a generated label in the format specified by `labelFormat` when the `autoLabel` option is `always` or `dev-only`. They _will_ include labels manually added with the Emotion’s custom `label` property, but won’t label the style based on the variable name like `@emotion/babel-plugin`.
2. Labels including the `[filename]` segment include the file extension, breaking selectors. e.g. if a style is defined in a file called `styles.js`, it will generate a className including `styles.js`, which breaks the CSS selector since it includes an illegal `.` character.

### Expected Behavior

1. `className`s add via the `css` prop should be labelled the same way as `className`s generated via the `@emotion/styled` `styled` function.
2. Labels including the `[filename]` segment should exclude the filename’s extension.

### To Reproduce

Reproduction here: https://github.com/GrantASL19/nextjs-swc-emotion-labels-regression

First, start as-is with `npm run dev`.

Given the Home component:

```tsx
const Home = () => (
  <div>
    <Basic>Cool Styles</Basic>
    <Combined>
      With <code>:hover</code>.
    </Combined>
    <Animated animation={bounce}>Let's bounce.</Animated>

    <p css={paragraph}>Paragraph styled with css prop</p>
    <p css={paragraphWithLabel}>Paragraph styled with css prop including label declaration</p>
  </div>
)
```

The Babel output is:

```html
<div>
  <div class="css-wf1qef-customPrefix-styles_Basic-customPrefix-styles_basicStyles e1rufc602">
    Cool Styles
  </div>
  <div class="css-17m2vpo-customPrefix-styles_Combined-customPrefix-styles_basicStyles-customPrefix-styles_hoverStyles e1rufc601">
    With <code>:hover</code>.
  </div>
  <div class="css-1ldxtiy-customPrefix-styles_Animated-customPrefix-styles_basicStyles-customPrefix-styles_hoverStyles e1rufc600">
    Let's bounce.
  </div>
  <p class="css-m78a9o-customPrefix-pages_paragraph">Paragraph styled with css prop</p>
  <p class="css-mlztiv-paragraphWithLabel-customPrefix-pages_paragraphWithLabel">
    Paragraph styled with css prop including label declaration
  </p>
</div>
```

Next delete or rename `.babelrc` to opt out of Babel and enable the SWC transform, then restart the dev server with `npm run dev`.

The SWC output is:

```html
<div>
  <div class="css-1dn0apk-customPrefix-styles.js_Basic e1y5w0sl0">
    Cool Styles
  </div>
  <div class="css-7037ma-customPrefix-styles.js_Combined e1y5w0sl1">
    With <code>:hover</code>.
  </div>
  <div class="css-q0mp9y-customPrefix-styles.js_Animated e1y5w0sl2">
    Let's bounce.
  </div>
  <p class="css-1g4xn9m-Home">Paragraph styled with css prop</p>
  <p class="css-upkvlf-paragraphWithLabel">
    Paragraph styled with css prop including label declaration
  </p>
</div>
```

Note that the styles from `styles.js` will no longer work because of the invalid generated className, and the labels for the elements styled with the `css` prop won’t match the `@emotion/babel-plugin` output (e.g. `customPrefix` is gone, and only the element with `label: paragraphWithLabel;` has `paragraphWithLabel` included in its `className`).

---

FWIW I’ve never liked how verbose the `@emotion/babel-plugin` generated `className`s are, but we do heavily rely on the auto-labelling to identify which component and style variable is applied to an element in the browser devtools.

So e.g. an output like `css-m78a9o-customPrefix-pages_paragraph` would be totally fine, but `css-1g4xn9m-Home` is a regression because we can’t identify that the style comes from the `paragraph` variable.

Also we exclusively use the `css` prop (not `styled`), so I can’t say if the `styled` behaviour is expected or not.

**Comment by Brooooooklyn** [comment]
@GrantASL19 Unfortunately we don't intend to be completely consistent with the behavior of the babel plugin, because there are some behaviors and logic that are difficult to implement in Rust or that we think are very performance-compromising

**Comment by GrantASL19** [comment]
> Unfortunately we don't intend to be completely consistent with the behavior of the babel plugin, because there are some behaviors and logic that are difficult to implement in Rust or that we think are very performance-compromising

That makes sense, thanks.

Is there any chance of a transform that essentially does this (when `autoLabel` is enabled)?

Input:

```ts
const paragraph = css`
  color: orangered;
`;
```

Output:

```ts
const paragraph = css`
  label: paragraph;
  color: orangered;
`;
```

The details of Babel/SWC transforms are unfortunately out of my depth, but (aside from the filename bug) this would essentially cover my use case.

AFAIK Babel recursively includes the name of every variable interpolated in the `css` string, which is IMO overkill and can make the label very difficult to read. I can’t speak for all users but I think most would be okay with just a label derived from the variable name. Without that it’s much more difficult to look at rendered HTML and figure out where its style came from.

**Comment by MonstraG** [comment]
Was extremely confused by not seing any labels even after setting `autoLabel: always`, untill read this. We practically only use `styled()` and as a result approximetely 0% of the classes are labeled.

Adding `label: asdsad` inside `styled` works (by adding file name instead).

Can this be fixed?
